modules/snmp-dest/snmpdest.c

```diff
@@ -0,0 +1,899 @@
+                }
+              else
+                ADD_ARG("authNoPriv");
+            }
+        }
+    }
+  ADD_ARG("localhost");
+  ADD_ARG("42");
+  ADD_ARG("coldStart.0");
+  gint arg = snmp_parse_args(argc, args, &self->session, "C:", optProc);
+
+  if (arg == -1 || arg == -2)
+    {
+      for (i = 0; i < argc; ++i)
+        g_free(args[i]);
+      free_session(&self->session);
+      return FALSE;
+    }
+  /* set our own address - the parse_args uses 'localhost' */
+  self->session.peername = self->host_port->str;
+
+  self->session.callback = snmp_input;
+  self->session.callback_magic = NULL;
+
+  /* this block is copied from the snmptrap.c - necessary? */
+  if (self->session.version == SNMP_VERSION_3)
+    {
+      /*
+       * for traps, we use ourselves as the authoritative engine
+       * which is really stupid since command line apps don't have a
+       * notion of a persistent engine.  Hence, our boots and time
+       * values are probably always really wacked with respect to what
+       * a manager would like to see.
+       *
+       * The following should be enough to:
+       *
+       * 1) prevent the library from doing discovery for engineid & time.
+       * 2) use our engineid instead of the remote engineid for
+       * authoritative & privacy related operations.
+       * 3) The remote engine must be configured with users for our engineID.
+       *
+       * -- Wes
+       */
+
+      /*
+       * setup the engineID based on IP addr.  Need a different
+       * algorthim here.  This will cause problems with agents on the
+       * same machine sending traps.
+       */
+      setup_engineID(NULL, NULL);
+
+      /*
+       * pick our own engineID
+       */
+      if (self->session.securityEngineIDLen == 0 ||
+          self->session.securityEngineID == NULL)
+        {
+          self->session.securityEngineID =
+            snmpv3_generate_engineID(&self->session.securityEngineIDLen);
+        }
+      if (self->session.contextEngineIDLen == 0 ||
+          self->session.contextEngineID == NULL)
+        {
+          self->session.contextEngineID =
+            snmpv3_generate_engineID(&self->session.contextEngineIDLen);
+        }
+
+      /*
+       * set boots and time, which will cause problems if this
+       * machine ever reboots and a remote trap receiver has cached our
+       * boots and time...  I'll cause a not-in-time-window report to
+       * be sent back to this machine.
+       */
+      if (self->session.engineBoots == 0)
+        self->session.engineBoots = 1;
+      if (self->session.engineTime == 0)    /* not really correct, */
+        self->session.engineTime = get_uptime();  /* but it'll work. Sort of. */
+    }
+
+  self->ss = snmp_add(&self->session,
+                      netsnmp_transport_open_client("snmptrap", self->session.peername),
+                      NULL, NULL);
+  if (self->ss == NULL)
+    {
+      for (i = 0; i < argc; ++i)
```

Comment:
This block is used two times? Just a suggestion, but you can avoid duplication by a goto error rollback method.

Reply:
done